man/jk_procmailwrapper.8
.TH jk_procmailwrapper 8 17-07-2007 JAILKIT jk_procmailwrapper

.SH NAME
jk_procmailwrapper \- a wrapper to prevent procmail execution by jailed users

.SH DESCRIPTION

jk_procmailwrapper is a wrapper around the regular procmail utility. For regular users it will execute the normal procmail utility. For users that are in a chroot jail (the home directory contains a . character) it will change root into the jail, and execute the procmail utility inside the jail. This enables safe mail delivery for jailed users.

Without this utility, users in a jail can execute commands outside the jail by placing them in their .procmailrc.

It should be used as a replacement for procmail in your mail server's configuration.

.SH WARNING
If your mailserver allows users to have a .forward file, they can still run scripts outside the jail! Make sure users cannot use a .forward file to run scripts on the real system!!

.SH "SEE ALSO"
.BR jailkit(8)
.BR jk_check(8)
.BR jk_chrootlaunch(8)
.BR jk_chrootsh(8)
.BR jk_cp(8)
.BR jk_init(8)
.BR jk_jailuser(8)
.BR jk_list(8)
.BR jk_lsh(8)
.BR jk_socketd(8)
.BR jk_uchroot(8)
.BR jk_update(8)
.BR chroot(2)
.BR procmail(1)

.SH COPYRIGHT

Copyright (C) 2003, 2004, 2005, 2006, 2007, Olivier Sessink

Copying and distribution of this file, with or without modification,
are permitted in any medium without royalty provided the copyright
notice and this notice are preserved.


debian/changelog
jailkit (2.4-1) unstable; urgency=low

  * added a new jk_uchroot utility
  * fixed a rare crash in configuration file parsing 
  * added new jk_init configuration file options
  * added new options to jk_cp and jk_init to retain ownership when copying files into the jail
  * improved the documentation
  * improved several chroot security checks 
  * fixed some build issues on OpenBSD
  * improved error reporting

 -- Olivier Sessink <olivier@cort.fakenet>  Tue, 17 Jul 2007 20:32:41 +0200

jailkit (2.3-1) unstable; urgency=low

  * various small fixes for the new jk_update utility
  * added a -k option to jk_cp, jk_init and jk_update to use hardlinks instead of copying
  * various fixes to install jailkit in a different location

 -- Olivier Sessink <olivier@aria.fakenet>  Thu, 28 Dec 2006 13:51:00 +0100

jailkit (2.2-1) unstable; urgency=low

  * Jailkit 2.2 released with a new utility jk_update to keep jails up-to-date
    and various improvements on *BSD operating systems

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Fri, 8 Dec 2006 00:00:00 +0200

jailkit (2.1-2) unstable; urgency=low

  * tarballs are uploaded again, there was a Makefile issue

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Fri, 8 Sep 2006 00:00:00

jailkit (2.1-1) unstable; urgency=low

  * improved logging
  * allowed sub-second intervals for jk_socketd
  * improved various checks in jk_jailuser
  * improved handling of recursive directories in jk_cp
  * new utility jk_list to list all jailed processes

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Thu, 7 Sep 2006 00:00:00

jailkit (2.0-1) unstable; urgency=low

  * fixes to allow jk_chrootsh to function on systems that use a name service
    caching daemon, or ldap authentication and others
  * some options for jk_chrootsh to relax some strict checks
  * replaced jk_addjailuser with jk_jailuser, which should work on more
    platforms
  * added the DEFAULT section to jk_lsh.ini and jk_chrootsh.ini to allow
    easier configuration

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Wed, 9 Aug 2006 00:00:00

jailkit (1.3-1) unstable; urgency=low

  * fixes some issues with Gentoo Linux
  * fixes some FreeBSD issues
  * fixes option --force for jk_init
  * better defaults in jk_init.ini
  * much more verbose error reporting by jk_lsh and jk_chrootsh

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Thu,  5 Aug 2004 19:22:41

jailkit (1.2-1) unstable; urgency=low

  * new upstream 1.2 release
  * mainly *BSD and MacOSX build related fixes

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Thu, 26 Aug 2004 21:34:01 +0200

jailkit (1.1-2) unstable; urgency=low

  * adding lintian override file, package now is lintian clean

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Sun, 15 Aug 2004 11:47:44 +0200

jailkit (1.1-1) unstable; urgency=low

  * new upstream 1.1 release
  * fixes some leftovers from /usr/bin/ to /usr/sbin/ transition
  * fixes minor bug in jk_init

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Fri, 13 Aug 2004 21:23:43 +0200

jailkit (1.0-1) unstable; urgency=low

  * new upstream 1.0 release
  * jk_adduser improvements
  * documentation improvements
  * moved all binaries to /usr/sbin/

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Thu,  5 Aug 2004 19:22:41 +0200

jailkit (0.9-1) unstable; urgency=low

  * new upstream 0.9 release
  * jk_cp and jk_init speed improvements

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Sun, 23 May 2004 21:28:17 +0200

jailkit (0.8-1) unstable; urgency=low

  * new upstream 0.8 release
  * adds wildcard expansion (*,?,~, etc.) to jk_lsh

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Tue, 18 May 2004 21:05:08 +0200

jailkit (0.7-1) unstable; urgency=low

  * Initial Release.
  * First Debian package

 -- Olivier Sessink <olivier@lx.student.wau.nl>  Sat,  1 May 2004 12:02:18 +0200



src/jk_procmailwrapper.c
/*
 *
 * jk_procmailwrapper
 * this program will simply execute procmail for users that are not in a jail
 * and it will exit() for users that are in a jail (mail will *not* be delivered)
 *
 * this will probably extended in the near future
 *
Copyright (c) 2003, 2004, 2005, Olivier Sessink
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions 
are met:
  * Redistributions of source code must retain the above copyright 
    notice, this list of conditions and the following disclaimer.
  * Redistributions in binary form must reproduce the above 
    copyright notice, this list of conditions and the following 
    disclaimer in the documentation and/or other materials provided 
    with the distribution.
  * The names of its contributors may not be used to endorse or 
    promote products derived from this software without specific 
    prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
POSSIBILITY OF SUCH DAMAGE.
*/
/* #define DEBUG */

#include "config.h"

#include <string.h>
#include <stdio.h>
#include <pwd.h>
#include <sys/types.h>
#include <unistd.h>
#include <stdlib.h>
#include <syslog.h>
#include <errno.h>
#include <grp.h>

#include "jk_lib.h"

#define PROGRAMNAME "jk_procmailwrapper"

int user_is_chrooted(const char *homedir) {
	char *tmp;
	tmp = strchr(homedir, '.');
	if (tmp != NULL) {
		return 1;
	}
	return 0;
}

void clean_exit(char * name, int error) {
	printf("%s, exiting with error %d\n", name, error);
	exit(error);
}

int main (int argc, char **argv, char **envp) {
	int i;
	struct passwd *pw=NULL;
	struct group *gr=NULL;
	char *jaildir=NULL, *newhome=NULL;

	DEBUG_MSG(PROGRAMNAME", started\n");
	pw = getpwuid(getuid());
	if (!user_is_chrooted(pw->pw_dir)) {
		/* if the user does not have a chroot homedir, we start the normal procmail now,
		but first we drop all privileges */
		if (setgid(getgid())) {
			syslog(LOG_ERR, "abort, failed to become gid %d", getgid());
			exit(34);
		}
		if (initgroups(pw->pw_name, getgid())) {
			syslog(LOG_ERR, "abort, failed to initgroups for user %s and group %d", pw->pw_name, getgid());
			exit(35);
		}
		if (setuid(getuid())) {
			syslog(LOG_ERR, "abort, failed to become uid %d", getuid());
			exit(36);
		}
		execve(PROCMAILPATH, argv, envp);
		/* if we get here, there is something wrong */
		exit(1);
	}
	/* OK, so the user is a jailed user, now we start checking things!! */
	
	/* open the log facility */
	openlog(PROGRAMNAME, LOG_PID, LOG_AUTH);
	
	/* check if it us that the user wants */
	{
		char *tmp = strrchr(argv[0], '/');
		if (!tmp) {
			tmp = argv[0];
		} else {
			tmp++;
		}
		if (strcmp(tmp, PROGRAMNAME) && (tmp[0] != '-' || strcmp(&tmp[1], PROGRAMNAME))) {
			DEBUG_MSG("wrong name, tmp=%s, &tmp[1]=%s\n", tmp, &tmp[1]);
			syslog(LOG_ERR, "abort, "PROGRAMNAME" is called as %s", argv[0]);
			exit(1);
		}
	}
	DEBUG_MSG("close filedescriptors\n");
	/* open file descriptors can be used to break out of a chroot, so we close all of them, except for stdin,stdout and stderr */
	for (i=getdtablesize();i>3;i--) {
		while (close(i) != 0 && errno == EINTR);
	}

	/* now test if we are setuid root (the effective user id must be 0, and the real user id > 0 */
	if (geteuid() != 0 || getuid() == 0) {
		syslog(LOG_ERR, "abort, "PROGRAMNAME" is not setuid root, or is run by root");
		exit(11);
	}

	DEBUG_MSG("get user info\n");
	gr = getgrgid(getgid());
	if (!pw || !gr) {
		syslog(LOG_ERR, "abort, failed to get user or group information for %d:%d", getuid(), getgid());
		exit(13);
	}

	/* now we clear the environment */
	clearenv();

	if (pw->pw_gid != getgid()) {
		syslog(LOG_ERR, "abort, the group ID from /etc/passwd (%d) does not match the group ID we run with (%d)", pw->pw_gid, getgid());
		exit(15);
	}
	if (!pw->pw_dir || strlen(pw->pw_dir) ==0 || strchr(pw->pw_dir, '.') == NULL) {
		syslog(LOG_ERR, "abort, the homedir does not contain the jail pattern path/./path");
		exit(17);
	}
	DEBUG_MSG("get jaildir\n");
	if (!getjaildir(pw->pw_dir, &jaildir, &newhome)) {
		syslog(LOG_ERR, "abort, failed to read the jail and the home from %s",pw->pw_dir);
		exit(17);
	}
	DEBUG_MSG("get chdir()\n");
	if (chdir(jaildir) != 0) {
		syslog(LOG_ERR, "abort, failed to chdir() to %s",jaildir);
		exit(19);
	} else {
		/* test if it really succeeded */
		char *test = get_current_dir_name();
		if (strcmp(jaildir, test) != 0) {
			syslog(LOG_ERR, "abort, current dir != %s after chdir()",jaildir);
			exit(21);
		}
		free(test);
	}		
	
	/* here do test the ownership of the jail and the homedir and such
	the function testsafepath doe exit itself on any failure */
	{ 
		int ret;
		DEBUG_MSG("test paths\n");
		ret = testsafepath(jaildir,0,0);
		if (ret != 0) {
			syslog(LOG_ERR, "abort, path %s is not a safe jail, check ownership and permissions", jaildir);
			exit(53);	
		}
		ret = testsafepath(pw->pw_dir, getuid(), getgid());
		if ((ret & TESTPATH_NOREGPATH) ) {
			syslog(LOG_ERR, "abort, path %s is not a directory", pw->pw_dir);
			exit(53);	
		}
		if ((ret & TESTPATH_OWNER) ) {
			syslog(LOG_ERR, "abort, path %s is not owned by %d", pw->pw_dir,getuid());
			exit(53);
		}
	}
	/* do a final log message */
	syslog(LOG_INFO, "now entering jail %s for user %d", jaildir, getuid());
	
	DEBUG_MSG("chroot()\n");
	/* do the chroot() call */
	if (chroot(jaildir)) {
		syslog(LOG_ERR, "abort, failed to chroot() to %s", jaildir);
		exit(33);
	}
	
	/* drop all privileges, it seems that we first have to setgid(), 
		then we have to call initgroups(), 
		then we call setuid() */
	if (setgid(getgid())) {
		syslog(LOG_ERR, "abort, failed to become gid %d", getgid());
		exit(34);
	}
	if (initgroups(pw->pw_name, getgid())) {
		syslog(LOG_ERR, "abort, failed to initgroups for user %s and group %d", pw->pw_name, getgid());
		exit(35);
	}
	if (setuid(getuid())) {
		syslog(LOG_ERR, "abort, failed to become uid %d", getuid());
		exit(36);
	}
	
	/* test for user and group info, is it the same? checks username, groupname and home */
	{
		char *oldpw_name,*oldgr_name;
		oldpw_name = strdup(pw->pw_name);
		oldgr_name = strdup(gr->gr_name);
		
		pw = getpwuid(getuid());
		gr = getgrgid(getgid());
		if (!pw || !gr) {
			syslog(LOG_ERR, "abort, failed to get user and group information in the jail for %d:%d", getuid(), getgid());
			exit(35);
		}
		if (strcmp(pw->pw_name, oldpw_name)!=0 || strcmp(gr->gr_name, oldgr_name)!=0) {
			syslog(LOG_ERR, "abort, user or group names differ inside the jail for %d:%d", getuid(), getgid());
			exit(37);
		}
		if (strcmp(pw->pw_dir, newhome)!=0) {
			syslog(LOG_ERR, "abort, home directory is incorrect inside the jail for %d:%d", getuid(), getgid());
			exit(39);
		}
		free(oldpw_name);
		free(oldgr_name);
	}
	
	/* test procmail in the jail, it is not allowed to be setuid() or setgid()
	it is common to have procmail setuid() root and setgid() mail in the regular 
	system, but it is for most situations not required, and therefore very much 
	not recommended inside a jail. So we will simply exit because it is a 
	security risk */
	testsafepath(PROCMAILPATH,0,0);
	
	/* prepare the new environment */
	setenv("HOME",newhome,1);
	setenv("USER",pw->pw_name,1);
	if (chdir(newhome) != 0) {
		syslog(LOG_ERR, "abort, failed to chdir() inside the jail to %s",newhome);
		exit(41);
	}

	/* cleanup before execution */
	free(newhome);
	free(jaildir);

	/* now execute the jailed shell */
	/*execl(pw->pw_shell, pw->pw_shell, NULL);*/
	{
		char **newargv;
		int i;
		newargv = malloc0((argc+1)*sizeof(char *));
		newargv[0] = PROCMAILPATH;
		for (i=1;i<argc;i++) {
			newargv[i] = argv[i];
		}
		execv(PROCMAILPATH, newargv);
	}
	DEBUG_MSG(strerror(errno));
	syslog(LOG_ERR, "WARNING: could not execute %s for user %d:%d",PROCMAILPATH,getuid(),getgid());
	
	exit(111);
}
